README.md
[![Build Status](https://travis-ci.org/mmisw/orr-ont.svg?branch=master)](https://travis-ci.org/mmisw/orr-ont)
[![Coverage Status](https://coveralls.io/repos/github/mmisw/orr-ont/badge.svg?branch=master)](https://coveralls.io/github/mmisw/orr-ont?branch=master)


The Ontology Registry and Repository (ORR) integrates semantic web services and tools
created by the Marine Metadata Interoperability project, MMI, toward the realization
of its [Semantic Framework](http://marinemetadata.org/semanticframework) vision.

[![](http://marinemetadata.org/files/mmi/semantic_framework.jpg)](http://marinemetadata.org/semanticframework)

Current funding is provided by the U.S. National Science Foundation through the
[Cross-Domain Observational Metadata for Environmental Sensing (X-DOMES) project](
https://www.earthcube.org/group/x-domes).

# What's orr-ont?

The ORR system is comprised of:

| component | description |
|-----------|-------------|
| [orr-ont](https://github.com/mmisw/orr-ont)       | Backend / REST endpoint |
| [orr-portal](https://github.com/mmisw/orr-portal) | Front-end |

This is the repository for the **orr-ont** component.


# Deployment

Interested in having an ORR instance on your server? See
[http://mmisw.org/orrdoc/install/](http://mmisw.org/orrdoc/install/).

# Development

## Testing

Have a mongo service running on your machine, for example, on a separate terminal session:

    docker run -it --rm -p 27017:27017 mongo

Run the tests:

    sbt test


ChangeLog.md
## change log ##

* 2021-12-07:  3.8.8

- move sendMessage to facilitate debugging
- add debug logging related with email notifications

* 2020-07-02:  3.8.6

- for more flexible webapp configuration:
  if available, concatenate `/etc/orront/orrportal.config.js` contents into
  generated `js/local.config.js` for the webapp.

* 2020-05-29:  3.8.5

- include status and visibility in email notification about ontology registration

* 2020-04-15:  3.8.4

- resolve #63 "remove requirement for format parameter".
  In general, the `format` parameter will still be required if the format
  cannot be determined from other submitted information, for example, from
  the file extension in an upload.
  NOTE: The swagger spec for the API Doc to be updated (separate repo).

* 2019-08-27:  3.8.3

- set `User-Agent:ORR-Ont/x.y.z` in requests to triple-store
  as an attempt to avoid/alleviate the issue of middleware (like Cloudflare)
  complaining about the `User-Agent:Java/1.8..`

* 2019-01-18:  3.8.2

  - resolve #69 "Redirect to portal if HTML is requested"
    - minor refact: move `getRequestedFormat` up to BaseOntController
      from SelfHostedOntController for eventual reuse by OntController
    - OntController: do redirection to portal if `format=html` given or
      `Accept: *text/html*` header included

* 2019-01-15:  3.8.1

  - TODO expose relevant configuration to client

  - resolve #68 "Compressed responses".
    Actually scalatra 2.3.0 does seem to support gzip
    http://scalatra.org//guides/2.3/http/gzip.html
    (btw, upgrading scalatra will certainly require significant effort;
    I just gave it a quick try and many compile errors arise)

* 2019-01-15:  3.8.0
  - resolve https://github.com/mmisw/orr-portal/issues/126
    - capture size limit in new config entry `files.maxUploadFileSize`
      with default 150MB
    - add test for the file size limit

* 2017-11-22:  3.7.0
  - simplify configurability for the front-end, mainly when such front-end
    is incorporated as part of the installable ORR package.
    The strategy is to capture UI relevant config parameters in `orront.conf`,
    and from there automatically generate `local.config.js` (only when it's no
    already included to allow external adjustments there for some reason).

* 2017-11-17:  3.7.0
  - configuration files (`orront.conf` and `notifyemails`) now expected to be under
    a common base directory `/etc/orront/`.
    This base directory specified via "configDir" parameter in `application.conf`.
    The names `orront.conf` and `notifyemails` are now hard-coded.
    The common base directory simplifies things, in particular, it facilitates
    Docker volume mapping while also fixing issue of actually loading any runtime
    changes to `notifyemails` whenever admin notifications are to be sent (as
    originally intended).
    TODO documentation about the files under `/etc/orront/`.

* 2017-11-03:  3.6.8
  - if any, admin email notifications now queued and dispatched every 5min.
    (note that regular user-specific emails continue to be sent right away.)
  - refactor notification logic as preparation for better handling

* 2017-10-12:  3.6.7
  - handle self-resolution to complete #55
  - for recognized file extension need to differentiate between ontology and term requests

* 2017-10-11:  3.6.7
  - as synonym of `format` parameter also accept `_format`
  - resolve #55 "any additional mechanisms for format/serialization resolution?"
    - implement "file type extension" for ontology request
    - implement "file type extension" for term request.
      Order of precedence to determine desired format:
        1. 'format' parameter
        2. "file extension"
        3. Accept header
      Note: in contrast to the *ontology request* case, here we immediately try
      resolution of the uri excluding the file extension in the case of absent
      'format' parameter' and uri with file extension.
      Part of the reason is that it's a bit tricky to determine whether the
      resolution with the file extension is actually successful or not in the sense
      of the term actually having triples associated or simply being non existing.

* 2017-10-04:  3.6.6
  - for ttl term request, also handle explicit accept header "text/turtle"
  - use 'construct' query for term request with 'ttl' as desired format.
    This is, 'text/turtle' as accept header against triple store
    (not application/sparql-results+ttl)
  - preference for "ttl" over "n3" in fileTypesForRecognition

* 2017-09-20:  3.6.5
  - fix #54 "enable "ttl" as a stored format"
  - `test/rtesources/ont1.rdf`: add some blank node, mainly to inspect
    that nesting is used in ttl and trig (per jena capabilities)

* 2017-07-21:  3.6.4
  - fix #52 "ontology self-resolution handled as term request".
    SelfHostedOntController.resolve was using `request.getRequestURL.toString`
    for the uri instead of `setup.cfg.deployment.url + pathInfo`

* 2017-06-06:  3.6.3
  - `GET /term`
    - also accept and process `offset` parameter
    - a bit bigger default value (30) for `limit` parameter

* 2017-06-03:  3.6.2
  - resolve #50 "term search and other queries"
    - `containing=string[&in=spo]` in subject, `in=s`, by default
    - `predicate=skos:relatedMatch&subject=termIri`
    - `predicate=owl:sameAs&subject=termIri`
    - `predicate=rdfs:isDefinedBy&subject=termIri`
    - `predicate=rdf:type&subject=termIri`
    - `predicate=predicateURI&subject=termIri`
      (full predicate URI assumed when not recognized by prefix)
    - all of above can have `&limit=number` (default limit, 10)
  - examples:
    - https://mmisw.org/ont/api/v0/term?containing=sensor&in=s&limit=2
    - https://mmisw.org/ont/api/v0/term?predicate=skos:exactMatch&subject=http://mmisw.org/ont/mmi/platform/AirAndOuterSpaceBasedPlatform
    - https://mmisw.org/ont/api/v0/term?predicate=skos:relatedMatch&subject=http://mmisw.org/ont/ioos/parameter/air_temperature

src/main/scala/org/mmisw/orr/ont/Cfg.scala
// generated by tscfg 0.9.5 on Tue Jan 15 09:10:16 PST 2019
// source: src/main/resources/orront.spec.conf

package org.mmisw.orr.ont

case class Cfg(
  admin           : Cfg.Admin,
  agraph          : Cfg.Agraph,
  auth            : Cfg.Auth,
  branding        : Cfg.Branding,
  deployment      : Cfg.Deployment,
  email           : Cfg.Email,
  files           : Cfg.Files,
  googleAnalytics : scala.Option[Cfg.GoogleAnalytics],
  mongo           : Cfg.Mongo,
  recaptcha       : scala.Option[Cfg.Recaptcha]
)
object Cfg {
  case class Admin(
    email    : java.lang.String,
    extra    : scala.Option[java.lang.String],
    password : java.lang.String
  )
  object Admin {
    def apply(c: com.typesafe.config.Config): Cfg.Admin = {
      Cfg.Admin(
        email    = c.getString("email"),
        extra    = if(c.hasPathOrNull("extra")) Some(c.getString("extra")) else None,
        password = c.getString("password")
      )
    }
  }
        
  case class Agraph(
    host           : java.lang.String,
    initDelay      : scala.Option[scala.Long],
    password       : java.lang.String,
    port           : scala.Int,
    repoName       : java.lang.String,
    sparqlEndpoint : java.lang.String,
    userName       : java.lang.String
  )
  object Agraph {
    def apply(c: com.typesafe.config.Config): Cfg.Agraph = {
      Cfg.Agraph(
        host           = if(c.hasPathOrNull("host")) c.getString("host") else "localhost",
        initDelay      = if(c.hasPathOrNull("initDelay")) Some(c.getDuration("initDelay", java.util.concurrent.TimeUnit.SECONDS)) else None,
        password       = c.getString("password"),
        port           = if(c.hasPathOrNull("port")) c.getInt("port") else 10035,
        repoName       = if(c.hasPathOrNull("repoName")) c.getString("repoName") else "mmiorr",
        sparqlEndpoint = c.getString("sparqlEndpoint"),
        userName       = c.getString("userName")
      )
    }
  }
        
  case class Auth(
    secret : java.lang.String
  )
  object Auth {
    def apply(c: com.typesafe.config.Config): Cfg.Auth = {
      Cfg.Auth(
        secret = c.getString("secret")
      )
    }
  }
        
  case class Branding(
    contactUs    : scala.Option[java.lang.String],
    footer       : scala.Option[java.lang.String],
    instanceName : java.lang.String,
    logo         : scala.Option[java.lang.String],
    tou          : scala.Option[java.lang.String]
  )
  object Branding {
    def apply(c: com.typesafe.config.Config): Cfg.Branding = {
      Cfg.Branding(
        contactUs    = if(c.hasPathOrNull("contactUs")) Some(c.getString("contactUs")) else None,
        footer       = if(c.hasPathOrNull("footer")) Some(c.getString("footer")) else None,
        instanceName = c.getString("instanceName"),
        logo         = if(c.hasPathOrNull("logo")) Some(c.getString("logo")) else None,
        tou          = if(c.hasPathOrNull("tou")) Some(c.getString("tou")) else None
      )
    }
  }
        
  case class Deployment(
    sparqlEndpoint : java.lang.String,
    url            : java.lang.String
  )
  object Deployment {
    def apply(c: com.typesafe.config.Config): Cfg.Deployment = {
      Cfg.Deployment(
        sparqlEndpoint = c.getString("sparqlEndpoint"),
        url            = c.getString("url")
      )
    }
  }
        
  case class Email(
    account : Cfg.Email.Account,
    from    : java.lang.String,
    mailer  : java.lang.String,
    replyTo : java.lang.String,
    server  : Cfg.Email.Server
  )
  object Email {
    case class Account(
      password : java.lang.String,
      username : java.lang.String
    )
    object Account {
      def apply(c: com.typesafe.config.Config): Cfg.Email.Account = {
        Cfg.Email.Account(
          password = c.getString("password"),
          username = c.getString("username")
        )
      }
    }
          
    case class Server(
      debug : scala.Boolean,
      host  : java.lang.String,
      port  : scala.Int,
      prot  : java.lang.String
    )
    object Server {
      def apply(c: com.typesafe.config.Config): Cfg.Email.Server = {
        Cfg.Email.Server(
          debug = c.hasPathOrNull("debug") && c.getBoolean("debug"),
          host  = c.getString("host"),
          port  = c.getInt("port"),
          prot  = if(c.hasPathOrNull("prot")) c.getString("prot") else "smtps"
        )
      }
    }
          
    def apply(c: com.typesafe.config.Config): Cfg.Email = {
      Cfg.Email(
        account = Cfg.Email.Account(if(c.hasPathOrNull("account")) c.getConfig("account") else com.typesafe.config.ConfigFactory.parseString("account{}")),
        from    = c.getString("from"),
        mailer  = c.getString("mailer"),
        replyTo = c.getString("replyTo"),
        server  = Cfg.Email.Server(if(c.hasPathOrNull("server")) c.getConfig("server") else com.typesafe.config.ConfigFactory.parseString("server{}"))
      )
    }
  }
        
  case class Files(
    baseDirectory     : java.lang.String,
    maxUploadFileSize : scala.Long
  )
  object Files {
    def apply(c: com.typesafe.config.Config): Cfg.Files = {
      Cfg.Files(
        baseDirectory     = c.getString("baseDirectory"),
        maxUploadFileSize = if(c.hasPathOrNull("maxUploadFileSize")) c.getBytes("maxUploadFileSize") else 157286400
      )
    }
  }
        
  case class GoogleAnalytics(
    propertyId : java.lang.String
  )
  object GoogleAnalytics {
    def apply(c: com.typesafe.config.Config): Cfg.GoogleAnalytics = {
      Cfg.GoogleAnalytics(
        propertyId = c.getString("propertyId")
      )
    }
  }
        
  case class Mongo(
    db            : java.lang.String,
    host          : java.lang.String,
    ontologies    : java.lang.String,
    organizations : java.lang.String,
    port          : scala.Int,
    pw            : scala.Option[java.lang.String],
    user          : scala.Option[java.lang.String],
    users         : java.lang.String
  )
  object Mongo {
    def apply(c: com.typesafe.config.Config): Cfg.Mongo = {
      Cfg.Mongo(
        db            = if(c.hasPathOrNull("db")) c.getString("db") else "orr-ont",
        host          = if(c.hasPathOrNull("host")) c.getString("host") else "localhost",
        ontologies    = if(c.hasPathOrNull("ontologies")) c.getString("ontologies") else "ontologies",
        organizations = if(c.hasPathOrNull("organizations")) c.getString("organizations") else "organizations",
        port          = if(c.hasPathOrNull("port")) c.getInt("port") else 27017,
        pw            = if(c.hasPathOrNull("pw")) Some(c.getString("pw")) else None,
        user          = if(c.hasPathOrNull("user")) Some(c.getString("user")) else None,
        users         = if(c.hasPathOrNull("users")) c.getString("users") else "users"
      )
    }
  }
        
  case class Recaptcha(
    privateKey : java.lang.String,
    siteKey    : java.lang.String
  )
  object Recaptcha {
    def apply(c: com.typesafe.config.Config): Cfg.Recaptcha = {
      Cfg.Recaptcha(
        privateKey = c.getString("privateKey"),
        siteKey    = c.getString("siteKey")
      )
    }
  }
        
  def apply(c: com.typesafe.config.Config): Cfg = {
    Cfg(
      admin           = Cfg.Admin(if(c.hasPathOrNull("admin")) c.getConfig("admin") else com.typesafe.config.ConfigFactory.parseString("admin{}")),
      agraph          = Cfg.Agraph(if(c.hasPathOrNull("agraph")) c.getConfig("agraph") else com.typesafe.config.ConfigFactory.parseString("agraph{}")),
      auth            = Cfg.Auth(if(c.hasPathOrNull("auth")) c.getConfig("auth") else com.typesafe.config.ConfigFactory.parseString("auth{}")),
      branding        = Cfg.Branding(if(c.hasPathOrNull("branding")) c.getConfig("branding") else com.typesafe.config.ConfigFactory.parseString("branding{}")),
      deployment      = Cfg.Deployment(if(c.hasPathOrNull("deployment")) c.getConfig("deployment") else com.typesafe.config.ConfigFactory.parseString("deployment{}")),
      email           = Cfg.Email(if(c.hasPathOrNull("email")) c.getConfig("email") else com.typesafe.config.ConfigFactory.parseString("email{}")),
      files           = Cfg.Files(if(c.hasPathOrNull("files")) c.getConfig("files") else com.typesafe.config.ConfigFactory.parseString("files{}")),
      googleAnalytics = if(c.hasPathOrNull("googleAnalytics")) scala.Some(Cfg.GoogleAnalytics(c.getConfig("googleAnalytics"))) else None,
      mongo           = Cfg.Mongo(if(c.hasPathOrNull("mongo")) c.getConfig("mongo") else com.typesafe.config.ConfigFactory.parseString("mongo{}")),
      recaptcha       = if(c.hasPathOrNull("recaptcha")) scala.Some(Cfg.Recaptcha(c.getConfig("recaptcha"))) else None
    )
  }
}
      
